### ISSUE TITLE
memory leak in mz_compute_client::protocol::response::ComputeResponse>>::absorb_response

### ISSUE THREAD

**Issue by philip-stoev** [MaterializeInc/materialize#17719]
### What version of Materialize are you using?

9a03acba53ef65271f06713e8514999acae01b02

### How did you install Materialize?

Docker image

### What is the issue?

The `KafkaSourcesLarge` Zippy scenario, which does not use any user tables, eventually OOMs with environmentd consuming 15+ GB of memory.

I was able to reproduce this on a scratch machine. I was watching the memory usage for several hours, producing memory profiles and it was always < 1Gb with no discernible growth and then the next morning it had jumped significantly. The memory is credited to the stack trace shown below.

I am not sure when this started happening, but I am tagging this as a release blocker since it does not look good.

To reproduce, use the following scratch definition:

```
philip@materialize:~/materialize$ cat misc/scratch/ci-large.json
{
    "name": "c6a.12xlarge as used by the CI",
    "launch_script": "true",
    "instance_type": "c6a.12xlarge",
    "ami": "ami-09d56f8956ab235b3",
    "size_gb": 500,
    "tags": {}
}
```
and run

```
cd test/zippy
./mzcompose run default --scenario KafkaSourcesLarge --actions=1000000
```

and wait 12+ hours

### Relevant log output

```source
__clone
0x7f06679c0b43
std::sys::unix::thread::Thread::new::thread_start::h1541b4390bbc26ca
<alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hebda57b0236de612
<alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::ha93239d92e7b610d
<<std[3da7f01c8c2aa3e3]::thread::Builder>::spawn_unchecked_<<tokio[289416cb391c4ec3]::runtime::blocking::pool::Spawner>::spawn_thread::{closure#0}, ()>::{closure#1} as core[e12cfea94acfcb96]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
<std[3da7f01c8c2aa3e3]::thread::Builder>::spawn_unchecked_::<<tokio[289416cb391c4ec3]::runtime::blocking::pool::Spawner>::spawn_thread::{closure#0}, ()>::{closure#1}
std[3da7f01c8c2aa3e3]::panic::catch_unwind::<core[e12cfea94acfcb96]::panic::unwind_safe::AssertUnwindSafe<<std[3da7f01c8c2aa3e3]::thread::Builder>::spawn_unchecked_<<tokio[289416cb391c4ec3]::runtime::blocking::pool::Spawner>::spawn_thread::{closure#0}, ()>::{closure#1}::{closure#0}>, ()>
std[3da7f01c8c2aa3e3]::panicking::try::<(), core[e12cfea94acfcb96]::panic::unwind_safe::AssertUnwindSafe<<std[3da7f01c8c2aa3e3]::thread::Builder>::spawn_unchecked_<<tokio[289416cb391c4ec3]::runtime::blocking::pool::Spawner>::spawn_thread::{closure#0}, ()>::{closure#1}::{closure#0}>>
std[3da7f01c8c2aa3e3]::panicking::try::do_call::<core[e12cfea94acfcb96]::panic::unwind_safe::AssertUnwindSafe<<std[3da7f01c8c2aa3e3]::thread::Builder>::spawn_unchecked_<<tokio[289416cb391c4ec3]::runtime::blocking::pool::Spawner>::spawn_thread::{closure#0}, ()>::{closure#1}::{closure#0}>, ()>
<core[e12cfea94acfcb96]::panic::unwind_safe::AssertUnwindSafe<<std[3da7f01c8c2aa3e3]::thread::Builder>::spawn_unchecked_<<tokio[289416cb391c4ec3]::runtime::blocking::pool::Spawner>::spawn_thread::{closure#0}, ()>::{closure#1}::{closure#0}> as core[e12cfea94acfcb96]::ops::function::FnOnce<()>>::call_once
<std[3da7f01c8c2aa3e3]::thread::Builder>::spawn_unchecked_::<<tokio[289416cb391c4ec3]::runtime::blocking::pool::Spawner>::spawn_thread::{closure#0}, ()>::{closure#1}::{closure#0}
std[3da7f01c8c2aa3e3]::sys_common::backtrace::__rust_begin_short_backtrace::<<tokio[289416cb391c4ec3]::runtime::blocking::pool::Spawner>::spawn_thread::{closure#0}, ()>
<tokio[289416cb391c4ec3]::runtime::blocking::pool::Spawner>::spawn_thread::{closure#0}
<tokio[289416cb391c4ec3]::runtime::blocking::pool::Inner>::run
<tokio[289416cb391c4ec3]::runtime::blocking::pool::Task>::run
<tokio[289416cb391c4ec3]::runtime::task::UnownedTask<tokio[289416cb391c4ec3]::runtime::blocking::schedule::BlockingSchedule>>::run
<tokio[289416cb391c4ec3]::runtime::task::raw::RawTask>::poll
<tokio[289416cb391c4ec3]::runtime::task::harness::Harness<tokio[289416cb391c4ec3]::runtime::blocking::task::BlockingTask<<tokio[289416cb391c4ec3]::runtime::scheduler::multi_thread::worker::Launch>::launch::{closure#0}>, tokio[289416cb391c4ec3]::runtime::blocking::schedule::BlockingSchedule>>::poll
<tokio[289416cb391c4ec3]::runtime::task::harness::Harness<tokio[289416cb391c4ec3]::runtime::blocking::task::BlockingTask<<tokio[289416cb391c4ec3]::runtime::scheduler::multi_thread::worker::Launch>::launch::{closure#0}>, tokio[289416cb391c4ec3]::runtime::blocking::schedule::BlockingSchedule>>::poll_inner
tokio[289416cb391c4ec3]::runtime::task::harness::poll_future::<tokio[289416cb391c4ec3]::runtime::blocking::task::BlockingTask<<tokio[289416cb391c4ec3]::runtime::scheduler::multi_thread::worker::Launch>::launch::{closure#0}>, tokio[289416cb391c4ec3]::runtime::blocking::schedule::BlockingSchedule>
std[3da7f01c8c2aa3e3]::panic::catch_unwind::<core[e12cfea94acfcb96]::panic::unwind_safe::AssertUnwindSafe<tokio[289416cb391c4ec3]::runtime::task::harness::poll_future<tokio[289416cb391c4ec3]::runtime::blocking::task::BlockingTask<<tokio[289416cb391c4ec3]::runtime::scheduler::multi_thread::worker::Launch>::launch::{closure#0}>, tokio[289416cb391c4ec3]::runtime::blocking::schedule::BlockingSchedule>::{closure#0}>, core[e12cfea94acfcb96]::task::poll::Poll<()>>
std[3da7f01c8c2aa3e3]::panicking::try::<core[e12cfea94acfcb96]::task::poll::Poll<()>, core[e12cfea94acfcb96]::panic::unwind_safe::AssertUnwindSafe<tokio[289416cb391c4ec3]::runtime::task::harness::poll_future<tokio[289416cb391c4ec3]::runtime::blocking::task::BlockingTask<<tokio[289416cb391c4ec3]::runtime::scheduler::multi_thread::worker::Launch>::launch::{closure#0}>, tokio[289416cb391c4ec3]::runtime::blocking::schedule::BlockingSchedule>::{closure#0}>>
std[3da7f01c8c2aa3e3]::panicking::try::do_call::<core[e12cfea94acfcb96]::panic::unwind_safe::AssertUnwindSafe<tokio[289416cb391c4ec3]::runtime::task::harness::poll_future<tokio[289416cb391c4ec3]::runtime::blocking::task::BlockingTask<<tokio[289416cb391c4ec3]::runtime::scheduler::multi_thread::worker::Launch>::launch::{closure#0}>, tokio[289416cb391c4ec3]::runtime::blocking::schedule::BlockingSchedule>::{closure#0}>, core[e12cfea94acfcb96]::task::poll::Poll<()>>
<core[e12cfea94acfcb96]::panic::unwind_safe::AssertUnwindSafe<tokio[289416cb391c4ec3]::runtime::task::harness::poll_future<tokio[289416cb391c4ec3]::runtime::blocking::task::BlockingTask<<tokio[289416cb391c4ec3]::runtime::scheduler::multi_thread::worker::Launch>::launch::{closure#0}>, tokio[289416cb391c4ec3]::runtime::blocking::schedule::BlockingSchedule>::{closure#0}> as core[e12cfea94acfcb96]::ops::function::FnOnce<()>>::call_once
tokio[289416cb391c4ec3]::runtime::task::harness::poll_future::<tokio[289416cb391c4ec3]::runtime::blocking::task::BlockingTask<<tokio[289416cb391c4ec3]::runtime::scheduler::multi_thread::worker::Launch>::launch::{closure#0}>, tokio[289416cb391c4ec3]::runtime::blocking::schedule::BlockingSchedule>::{closure#0}
<tokio[289416cb391c4ec3]::runtime::task::core::Core<tokio[289416cb391c4ec3]::runtime::blocking::task::BlockingTask<<tokio[289416cb391c4ec3]::runtime::scheduler::multi_thread::worker::Launch>::launch::{closure#0}>, tokio[289416cb391c4ec3]::runtime::blocking::schedule::BlockingSchedule>>::poll
<tokio[289416cb391c4ec3]::loom::std::unsafe_cell::UnsafeCell<tokio[289416cb391c4ec3]::runtime::task::core::Stage<tokio[289416cb391c4ec3]::runtime::blocking::task::BlockingTask<<tokio[289416cb391c4ec3]::runtime::scheduler::multi_thread::worker::Launch>::launch::{closure#0}>>>>::with_mut::<core[e12cfea94acfcb96]::task::poll::Poll<()>, <tokio[289416cb391c4ec3]::runtime::task::core::Core<tokio[289416cb391c4ec3]::runtime::blocking::task::BlockingTask<<tokio[289416cb391c4ec3]::runtime::scheduler::multi_thread::worker::Launch>::launch::{closure#0}>, tokio[289416cb391c4ec3]::runtime::blocking::schedule::BlockingSchedule>>::poll::{closure#0}>
<tokio[289416cb391c4ec3]::runtime::task::core::Core<tokio[289416cb391c4ec3]::runtime::blocking::task::BlockingTask<<tokio[289416cb391c4ec3]::runtime::scheduler::multi_thread::worker::Launch>::launch::{closure#0}>, tokio[289416cb391c4ec3]::runtime::blocking::schedule::BlockingSchedule>>::poll::{closure#0}
<tokio[289416cb391c4ec3]::runtime::blocking::task::BlockingTask<<tokio[289416cb391c4ec3]::runtime::scheduler::multi_thread::worker::Launch>::launch::{closure#0}> as core[e12cfea94acfcb96]::future::future::Future>::poll
<tokio[289416cb391c4ec3]::runtime::scheduler::multi_thread::worker::Launch>::launch::{closure#0}
tokio[289416cb391c4ec3]::runtime::scheduler::multi_thread::worker::run
<tokio[289416cb391c4ec3]::macros::scoped_tls::ScopedKey<tokio[289416cb391c4ec3]::runtime::scheduler::multi_thread::worker::Context>>::set::<tokio[289416cb391c4ec3]::runtime::scheduler::multi_thread::worker::run::{closure#0}, ()>
tokio[289416cb391c4ec3]::runtime::scheduler::multi_thread::worker::run::{closure#0}
<tokio[289416cb391c4ec3]::runtime::scheduler::multi_thread::worker::Context>::run
<tokio[289416cb391c4ec3]::runtime::scheduler::multi_thread::worker::Context>::run_task
tokio[289416cb391c4ec3]::runtime::coop::budget::<core[e12cfea94acfcb96]::result::Result<alloc[7282c0069cd654ea]::boxed::Box<tokio[289416cb391c4ec3]::runtime::scheduler::multi_thread::worker::Core>, ()>, <tokio[289416cb391c4ec3]::runtime::scheduler::multi_thread::worker::Context>::run_task::{closure#0}>
tokio[289416cb391c4ec3]::runtime::coop::with_budget::<core[e12cfea94acfcb96]::result::Result<alloc[7282c0069cd654ea]::boxed::Box<tokio[289416cb391c4ec3]::runtime::scheduler::multi_thread::worker::Core>, ()>, <tokio[289416cb391c4ec3]::runtime::scheduler::multi_thread::worker::Context>::run_task::{closure#0}>
<tokio[289416cb391c4ec3]::runtime::scheduler::multi_thread::worker::Context>::run_task::{closure#0}
<tokio[289416cb391c4ec3]::runtime::task::LocalNotified<alloc[7282c0069cd654ea]::sync::Arc<tokio[289416cb391c4ec3]::runtime::scheduler::multi_thread::handle::Handle>>>::run
<tokio[289416cb391c4ec3]::runtime::task::raw::RawTask>::poll
<tokio[289416cb391c4ec3]::runtime::task::harness::Harness<<mz_compute_client[34def3868ff4a0f9]::controller::replica::ReplicaTask<mz_repr[919a3e299f41795b]::timestamp::Timestamp>>::run::{closure#0}, alloc[7282c0069cd654ea]::sync::Arc<tokio[289416cb391c4ec3]::runtime::scheduler::multi_thread::handle::Handle>>>::poll
<tokio[289416cb391c4ec3]::runtime::task::harness::Harness<<mz_compute_client[34def3868ff4a0f9]::controller::replica::ReplicaTask<mz_repr[919a3e299f41795b]::timestamp::Timestamp>>::run::{closure#0}, alloc[7282c0069cd654ea]::sync::Arc<tokio[289416cb391c4ec3]::runtime::scheduler::multi_thread::handle::Handle>>>::poll_inner
tokio[289416cb391c4ec3]::runtime::task::harness::poll_future::<<mz_compute_client[34def3868ff4a0f9]::controller::replica::ReplicaTask<mz_repr[919a3e299f41795b]::timestamp::Timestamp>>::run::{closure#0}, alloc[7282c0069cd654ea]::sync::Arc<tokio[289416cb391c4ec3]::runtime::scheduler::multi_thread::handle::Handle>>
std[3da7f01c8c2aa3e3]::panic::catch_unwind::<core[e12cfea94acfcb96]::panic::unwind_safe::AssertUnwindSafe<tokio[289416cb391c4ec3]::runtime::task::harness::poll_future<<mz_compute_client[34def3868ff4a0f9]::controller::replica::ReplicaTask<mz_repr[919a3e299f41795b]::timestamp::Timestamp>>::run::{closure#0}, alloc[7282c0069cd654ea]::sync::Arc<tokio[289416cb391c4ec3]::runtime::scheduler::multi_thread::handle::Handle>>::{closure#0}>, core[e12cfea94acfcb96]::task::poll::Poll<()>>
std[3da7f01c8c2aa3e3]::panicking::try::<core[e12cfea94acfcb96]::task::poll::Poll<()>, core[e12cfea94acfcb96]::panic::unwind_safe::AssertUnwindSafe<tokio[289416cb391c4ec3]::runtime::task::harness::poll_future<<mz_compute_client[34def3868ff4a0f9]::controller::replica::ReplicaTask<mz_repr[919a3e299f41795b]::timestamp::Timestamp>>::run::{closure#0}, alloc[7282c0069cd654ea]::sync::Arc<tokio[289416cb391c4ec3]::runtime::scheduler::multi_thread::handle::Handle>>::{closure#0}>>
std[3da7f01c8c2aa3e3]::panicking::try::do_call::<core[e12cfea94acfcb96]::panic::unwind_safe::AssertUnwindSafe<tokio[289416cb391c4ec3]::runtime::task::harness::poll_future<<mz_compute_client[34def3868ff4a0f9]::controller::replica::ReplicaTask<mz_repr[919a3e299f41795b]::timestamp::Timestamp>>::run::{closure#0}, alloc[7282c0069cd654ea]::sync::Arc<tokio[289416cb391c4ec3]::runtime::scheduler::multi_thread::handle::Handle>>::{closure#0}>, core[e12cfea94acfcb96]::task::poll::Poll<()>>
<core[e12cfea94acfcb96]::panic::unwind_safe::AssertUnwindSafe<tokio[289416cb391c4ec3]::runtime::task::harness::poll_future<<mz_compute_client[34def3868ff4a0f9]::controller::replica::ReplicaTask<mz_repr[919a3e299f41795b]::timestamp::Timestamp>>::run::{closure#0}, alloc[7282c0069cd654ea]::sync::Arc<tokio[289416cb391c4ec3]::runtime::scheduler::multi_thread::handle::Handle>>::{closure#0}> as core[e12cfea94acfcb96]::ops::function::FnOnce<()>>::call_once
tokio[289416cb391c4ec3]::runtime::task::harness::poll_future::<<mz_compute_client[34def3868ff4a0f9]::controller::replica::ReplicaTask<mz_repr[919a3e299f41795b]::timestamp::Timestamp>>::run::{closure#0}, alloc[7282c0069cd654ea]::sync::Arc<tokio[289416cb391c4ec3]::runtime::scheduler::multi_thread::handle::Handle>>::{closure#0}
<tokio[289416cb391c4ec3]::runtime::task::core::Core<<mz_compute_client[34def3868ff4a0f9]::controller::replica::ReplicaTask<mz_repr[919a3e299f41795b]::timestamp::Timestamp>>::run::{closure#0}, alloc[7282c0069cd654ea]::sync::Arc<tokio[289416cb391c4ec3]::runtime::scheduler::current_thread::Handle>>>::poll
<tokio[289416cb391c4ec3]::loom::std::unsafe_cell::UnsafeCell<tokio[289416cb391c4ec3]::runtime::task::core::Stage<<mz_compute_client[34def3868ff4a0f9]::controller::replica::ReplicaTask<mz_repr[919a3e299f41795b]::timestamp::Timestamp>>::run::{closure#0}>>>::with_mut::<core[e12cfea94acfcb96]::task::poll::Poll<()>, <tokio[289416cb391c4ec3]::runtime::task::core::Core<<mz_compute_client[34def3868ff4a0f9]::controller::replica::ReplicaTask<mz_repr[919a3e299f41795b]::timestamp::Timestamp>>::run::{closure#0}, alloc[7282c0069cd654ea]::sync::Arc<tokio[289416cb391c4ec3]::runtime::scheduler::current_thread::Handle>>>::poll::{closure#0}>
<tokio[289416cb391c4ec3]::runtime::task::core::Core<<mz_compute_client[34def3868ff4a0f9]::controller::replica::ReplicaTask<mz_repr[919a3e299f41795b]::timestamp::Timestamp>>::run::{closure#0}, alloc[7282c0069cd654ea]::sync::Arc<tokio[289416cb391c4ec3]::runtime::scheduler::current_thread::Handle>>>::poll::{closure#0}
<mz_compute_client[34def3868ff4a0f9]::controller::replica::ReplicaTask<mz_repr[919a3e299f41795b]::timestamp::Timestamp>>::run::{closure#0}
mz_compute_client[34def3868ff4a0f9]::controller::replica::run_message_loop::<mz_repr[919a3e299f41795b]::timestamp::Timestamp>::{closure#0}
<tokio[289416cb391c4ec3]::future::poll_fn::PollFn<mz_compute_client[34def3868ff4a0f9]::controller::replica::run_message_loop<mz_repr[919a3e299f41795b]::timestamp::Timestamp>::{closure#0}::{closure#1}> as core[e12cfea94acfcb96]::future::future::Future>::poll
mz_compute_client[34def3868ff4a0f9]::controller::replica::run_message_loop::<mz_repr[919a3e299f41795b]::timestamp::Timestamp>::{closure#0}::{closure#1}
<core[e12cfea94acfcb96]::pin::Pin<alloc[7282c0069cd654ea]::boxed::Box<dyn core[e12cfea94acfcb96]::future::future::Future<Output = core[e12cfea94acfcb96]::result::Result<core[e12cfea94acfcb96]::option::Option<mz_compute_client[34def3868ff4a0f9]::protocol::response::ComputeResponse>, anyhow[47fd497bd0648ee6]::Error>> + core[e12cfea94acfcb96]::marker::Send>> as core[e12cfea94acfcb96]::future::future::Future>::poll
<mz_service[74aa72eccbef3b27]::client::Partitioned<mz_service[74aa72eccbef3b27]::grpc::GrpcClient<mz_compute_client[34def3868ff4a0f9]::service::ComputeProtoServiceTypes>, mz_compute_client[34def3868ff4a0f9]::protocol::command::ComputeCommand, mz_compute_client[34def3868ff4a0f9]::protocol::response::ComputeResponse> as mz_service[74aa72eccbef3b27]::client::GenericClient<mz_compute_client[34def3868ff4a0f9]::protocol::command::ComputeCommand, mz_compute_client[34def3868ff4a0f9]::protocol::response::ComputeResponse>>::recv::{closure#0}
<mz_compute_client[34def3868ff4a0f9]::service::PartitionedComputeState<mz_repr[919a3e299f41795b]::timestamp::Timestamp> as mz_service[74aa72eccbef3b27]::client::PartitionedState<mz_compute_client[34def3868ff4a0f9]::protocol::command::ComputeCommand, mz_compute_client[34def3868ff4a0f9]::protocol::response::ComputeResponse>>::absorb_response
<alloc[7282c0069cd654ea]::vec::Vec<(mz_repr[919a3e299f41795b]::global_id::GlobalId, timely[93a8fc0b5165ed11]::progress::frontier::Antichain<mz_repr[919a3e299f41795b]::timestamp::Timestamp>)>>::push
<timely[93a8fc0b5165ed11]::progress::frontier::AntichainRef<mz_repr[919a3e299f41795b]::timestamp::Timestamp>>::to_owned
<alloc[7282c0069cd654ea]::raw_vec::RawVec<postgres_types[5bc169d189573a1e]::Field>>::reserve_for_push
```

**Comment by philip-stoev** [comment]
[environment-feb18.mzfg.zip](https://github.com/MaterializeInc/materialize/files/10773793/environment-feb18.mzfg.zip)

**Comment by antiguru** [comment]
Probably this line: https://github.com/MaterializeInc/materialize/blame/main/src/compute-client/src/service.rs#L279

This itself is likely not the cause of the problem but just a symptom.

**Comment by teskje** [comment]
@philip-stoev the "KafkaSourcesLarge" scenario contains Materialize restarts. Did you change the scenario to disable these or were you just lucky to catch a flame graph between restarts?

**Comment by philip-stoev** [comment]
@teskje yes what happened is that the test stopped making progress I think because a testdrive `kafka-ingest` step hung for some reason. So Mz was left to sit with a bunch of dataflows defined and no incoming kafka messages for 8+ hours. The test did not restart Mz any longer. This is how I found it in the morning and this allowed me to capture the flame graph.

**Comment by teskje** [comment]
Thanks Philip!

I have so far not been able to reproduce this with zippy. My initial run failed after three hours for an unknown reason, with nothing interesting in the memory profile.

However, I've looked at the code and I think @antiguru is spot-on. The line he pointed out is the only one that matches the provided stack trace and this is very likely just a symptom of another issue. Here is why:

* The `new_uppers` vector is the memory that gets leaked.
* This vector [is returned by `absorb_response` inside a `FrontierUppers` compute response](https://github.com/MaterializeInc/materialize/blob/03355126df2be212d52f369a86b7dc85d2a00d5d/src/compute-client/src/service.rs#L292).
* If we follow the call stack up, we see that that response [is received by the replica task](https://github.com/MaterializeInc/materialize/blob/03355126df2be212d52f369a86b7dc85d2a00d5d/src/compute-client/src/controller/replica.rs#L254).
* The replica task [pushes the response into its `response_tx`](https://github.com/MaterializeInc/materialize/blob/03355126df2be212d52f369a86b7dc85d2a00d5d/src/compute-client/src/controller/replica.rs#L258).
* The compute controller [pulls the response from the corresponding `response_rx`](https://github.com/MaterializeInc/materialize/blob/03355126df2be212d52f369a86b7dc85d2a00d5d/src/compute-client/src/controller/replica.rs#L114) and [propagates](https://github.com/MaterializeInc/materialize/blob/03355126df2be212d52f369a86b7dc85d2a00d5d/src/compute-client/src/controller/instance.rs#L315) [it](https://github.com/MaterializeInc/materialize/blob/03355126df2be212d52f369a86b7dc85d2a00d5d/src/compute-client/src/controller.rs#L343) [to](https://github.com/MaterializeInc/materialize/blob/03355126df2be212d52f369a86b7dc85d2a00d5d/src/compute-client/src/controller.rs#L350) [its](https://github.com/MaterializeInc/materialize/blob/03355126df2be212d52f369a86b7dc85d2a00d5d/src/compute-client/src/controller.rs#L589) [response handler](https://github.com/MaterializeInc/materialize/blob/03355126df2be212d52f369a86b7dc85d2a00d5d/src/compute-client/src/controller.rs#L591) (in a roundabout way, because we have to worry about cancelation safety).
* In the response handler, what was originally the `new_uppers` vector [gets deconstructed and turned into yet another vector](https://github.com/MaterializeInc/materialize/blob/03355126df2be212d52f369a86b7dc85d2a00d5d/src/compute-client/src/controller/instance.rs#L1083-L1086).

If I'm not mistaken, the memory profile shows us where a piece of memory was allocated, if the allocation is still alive. The response handler frees the `new_uppers` allocation, I think. Which means that the "leak" we see is caused by `FrontierUppers` responses never making it to the response handler in the compute controller.

The only place I can see that could accumulate `FrontierUppers` responses from `PartitionedComputeState` is the `response_tx`. This is an unbounded channel, so it accepts new responses until there is no memory left.

If my analysis is correct then that means the compute controller has stopped polling the `response_rx` (or it is polling slower than the replica task is pushing new responses). This is where it gets murky: The compute controller does not poll for new responses periodically, but only when asked to by the coordinator. So it is possible that the coordinator is somehow wedged and has stopped polling the compute controller. Of course it is equally possible for the compute controller to have wedged itself.

Here are two potential causes I can immediately come up with:
* The controller is waiting on CRDB in its critical path, blocking any progress.
* Something is pegging the CPU, slowing the receiving of messages from replicas down to a crawl.

Long story short: We probably don't have a memory leak but something else that degrades performance and causes messages to queue up. It's really hard to say what that is without a reliable repro.

Looking at our metrics, there is a memory leak on staging, but that exists on production as well and can be explained by https://github.com/MaterializeInc/materialize/issues/17741.

**Comment by philip-stoev** [comment]
I am now trying to reproduce on a scratch machine that I can then hand off to you. The issue has happened 5 times already (3 in CI and 2 in my testing).

**Comment by teskje** [comment]
I was able to reproduce this on the second try!

I think the issue is caused by CRDB connection issues. The last thing I see logged by the materialized container is a bunch of this:
```
environmentd: 2023-02-20T16:55:27.537090Z  WARN deadpool.postgres: Connection error: error communicating with the server: Connection reset by peer (os error 104)
environmentd: 2023-02-20T16:55:27.537099Z  WARN deadpool.postgres: Connection error: connection closed
environmentd: 2023-02-20T16:55:27.537150Z  WARN deadpool.postgres: Connection error: connection closed
environmentd: 2023-02-20T16:55:27.537160Z  WARN deadpool.postgres: Connection error: connection closed
environmentd: 2023-02-20T16:55:27.537200Z  WARN deadpool.postgres: Connection error: connection closed
environmentd: 2023-02-20T16:55:27.537209Z  WARN deadpool.postgres: Connection error: connection closed
environmentd: 2023-02-20T16:55:27.537225Z  WARN deadpool.postgres: Connection error: connection closed
environmentd: 2023-02-20T16:55:27.537237Z  WARN deadpool.postgres: Connection error: connection closed
environmentd: 2023-02-20T16:55:27.537265Z  WARN deadpool.postgres: Connection error: connection closed
environmentd: 2023-02-20T16:55:27.537284Z  WARN deadpool.postgres: Connection error: connection closed
cluster-u52-replica-54: 2023-02-20T16:55:27.537281Z  WARN deadpool.postgres: Connection error: connection closed
environmentd: 2023-02-20T16:55:27.537285Z  WARN deadpool.postgres: Connection error: connection closed
environmentd: 2023-02-20T16:55:27.537317Z  WARN deadpool.postgres: Connection error: connection closed
environmentd: 2023-02-20T16:55:27.537329Z  WARN deadpool.postgres: Connection error: connection closed
environmentd: 2023-02-20T16:55:27.537345Z  WARN deadpool.postgres: Connection error: connection closed
environmentd: 2023-02-20T16:55:27.537359Z ERROR mz_stash::postgres: postgres stash connection error: connection closed
environmentd: 2023-02-20T16:55:27.537398Z ERROR mz_stash::postgres: postgres stash connection error: connection closed
environmentd: 2023-02-20T16:55:27.537417Z  WARN deadpool.postgres: Connection error: connection closed
environmentd: 2023-02-20T16:55:27.537422Z ERROR mz_stash::postgres: postgres stash connection error: connection closed
environmentd: 2023-02-20T16:55:27.540699Z ERROR mz_stash::postgres: postgres stash connection error: connection closed
```

(It looks like the log level is WARN, so we are not seeing any logs at level INFO or below.)

I also see that the CRDB container is not running. It has crashed with:

```
*
* WARNING: disk slowness detected: unable to sync log files within 10s
*
*
* ERROR: Queued as error 1346fd1826a8418b8d22964bafd7e134
*
*
* FATAL: disk stall detected: unable to sync log files within 20s
*
```

So here is what I think is going on:
* CRDB crashes due the above fatal error (no idea what caused it).
* Adapter notices the loss of connection and retries forever, which makes it get stuck.
* Because of this, adapter stops calling the compute controllers `ready` method, which makes the compute controller stop pulling responses from the replicas' `response_rx`s.
* The replicas continue to send responses and the replica task continues to enqueue them in their `response_tx`s.
* The queued responses eventually take up all memory and crash environmentd.

This should be pretty easy to verify locally, by just shutting down CRDB and checking whether that makes adapter become stuck.

**Comment by teskje** [comment]
I'm able to verify this locally, using the patch below that makes the replica task print the queue size if it gets larger than 10. With that patch you can:
1. Run environmentd: `$ bin/environmentd --reset -- --log-filter=warn`, wait until it is ready.
2. Stop CRDB, e.g. (on Mac): `$ brew services stop cockroach`
3. Watch the output of environmentd for a while. Observe the replica response queues getting filled:
```
[...]
environmentd: 2023-02-21T10:25:07.197665Z ERROR mz_stash::postgres: postgres stash connection error: db error: ERROR: server is shutting down
environmentd: 2023-02-21T10:25:07.521230Z ERROR mz_stash::postgres: postgres stash connection error: db error: ERROR: server is shutting down
environmentd: 2023-02-21T10:25:07.538384Z ERROR mz_stash::postgres: postgres stash connection error: db error: ERROR: server is shutting down
environmentd: 2023-02-21T10:25:07.568542Z ERROR mz_stash::postgres: postgres stash connection error: db error: ERROR: server is shutting down
environmentd: 2023-02-21T10:25:15.204823Z  WARN mz_compute_client::controller::replica: response queue size: 10 replica_id=3
environmentd: 2023-02-21T10:25:15.280662Z  WARN mz_compute_client::controller::replica: response queue size: 10 replica_id=1
environmentd: 2023-02-21T10:25:15.289826Z  WARN mz_compute_client::controller::replica: response queue size: 10 replica_id=2
environmentd: 2023-02-21T10:25:20.239302Z  WARN mz_compute_client::controller::replica: response queue size: 20 replica_id=3
environmentd: 2023-02-21T10:25:20.308903Z  WARN mz_compute_client::controller::replica: response queue size: 20 replica_id=1
environmentd: 2023-02-21T10:25:20.317607Z  WARN mz_compute_client::controller::replica: response queue size: 20 replica_id=2
environmentd: 2023-02-21T10:25:25.275039Z  WARN mz_compute_client::controller::replica: response queue size: 30 replica_id=3
environmentd: 2023-02-21T10:25:25.324887Z  WARN mz_compute_client::controller::replica: response queue size: 30 replica_id=1
environmentd: 2023-02-21T10:25:25.345065Z  WARN mz_compute_client::controller::replica: response queue size: 30 replica_id=2
```

-----

```diff
diff --git a/src/compute-client/src/controller/replica.rs b/src/compute-client/src/controller/replica.rs
index 571640b6d..e427d7210 100644
--- a/src/compute-client/src/controller/replica.rs
+++ b/src/compute-client/src/controller/replica.rs
@@ -9,6 +9,8 @@

 //! A client for replicas of a compute instance.

+use std::sync::Arc;
+use std::sync::atomic::{AtomicUsize, Ordering};
 use std::time::Duration;

 use anyhow::bail;
@@ -53,6 +55,7 @@ pub(super) struct Replica<T> {
     /// If receiving from the channel returns `None`, the replica has failed
     /// and requires rehydration.
     response_rx: UnboundedReceiver<ComputeResponse<T>>,
+    queue_size: Arc<AtomicUsize>,
     /// A handle to the task that aborts it when the replica is dropped.
     _task: AbortOnDropHandle<()>,
     /// Configuration specific to this replica.
@@ -76,6 +79,7 @@ where
         // the replica.
         let (command_tx, command_rx) = unbounded_channel();
         let (response_tx, response_rx) = unbounded_channel();
+        let queue_size = Arc::new(AtomicUsize::new(0));

         let task = mz_ore::task::spawn(
             || format!("active-replication-replica-{id}"),
@@ -85,6 +89,7 @@ where
                 config: config.clone(),
                 command_rx,
                 response_tx,
+                queue_size: Arc::clone(&queue_size),
                 epoch,
                 metrics,
             }
@@ -94,6 +99,7 @@ where
         Self {
             command_tx,
             response_rx,
+            queue_size,
             _task: task.abort_on_drop(),
             config,
         }
@@ -111,7 +117,9 @@ where
     ///
     /// This method is cancellation safe.
     pub(super) async fn recv(&mut self) -> Option<ComputeResponse<T>> {
-        self.response_rx.recv().await
+        let resp = self.response_rx.recv().await;
+        self.queue_size.fetch_sub(1, Ordering::SeqCst);
+        resp
     }
 }

@@ -127,6 +135,7 @@ struct ReplicaTask<T> {
     command_rx: UnboundedReceiver<ComputeCommand<T>>,
     /// A channel upon which responses from the replica are delivered.
     response_tx: UnboundedSender<ComputeResponse<T>>,
+    queue_size: Arc<AtomicUsize>,
     /// A number (technically, pair of numbers) identifying this incarnation of the replica.
     /// The semantics of this don't matter, except that it must strictly increase.
     epoch: ClusterStartupEpoch,
@@ -147,6 +156,7 @@ where
             build_info,
             command_rx,
             response_tx,
+            queue_size,
             epoch,
             metrics,
         } = self;
@@ -170,6 +180,7 @@ where
             replica_id,
             command_rx,
             response_tx,
+            queue_size,
             build_info,
             addrs,
             cmd_spec,
@@ -195,6 +206,7 @@ async fn run_message_loop<T>(
     replica_id: ReplicaId,
     mut command_rx: UnboundedReceiver<ComputeCommand<T>>,
     response_tx: UnboundedSender<ComputeResponse<T>>,
+    queue_size: Arc<AtomicUsize>,
     build_info: &BuildInfo,
     addrs: Vec<String>,
     cmd_spec: CommandSpecialization,
@@ -259,6 +271,10 @@ where
                     // Controller is no longer interested in this replica. Shut down.
                     break;
                 }
+                let size = queue_size.fetch_add(1, Ordering::SeqCst) + 1;
+                if size % 10 == 0 {
+                    tracing::warn!(?replica_id, "response queue size: {size}");
+                }
             }
         }
     }
```

**Comment by teskje** [comment]
I'm assigning this back to the QA team since I don't think there is anything else compute can do here. It *seems* like the issue is that the disk cannot keep up with the load CRDB puts on it, though I don't know how to debug this further.

**Comment by philip-stoev** [comment]
Yes, I reproduced it again and I am able to see that crdb went down in this case as well ...
